### PR TITLE
Restrict fields submitted through Salesforce API

### PR DESCRIPTION
**Pull request by ahorner**
At times it is beneficial to pull down fields from Salesforce which we
cannot push back up to Salesforce — read-only fields or formulas may
sometimes contain information which is valuable to synchronize into the
database.

In order to prevent conflicts in these situations, we can consult the
Salesforce API directly for information regarding whether or not a field
is read-only, and parse down our attributes accordingly.

### CODE REVIEW

File: lib/restforce/db/field_processor.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  module DB
+
+    # Restforce::DB::FieldProcessor encapsulates logic for preventing
+    # information for unwriteable fields from being submitted to Salesforce.
+    class FieldProcessor
+
+      # Internal: Get a global cache with which to store/fetch the writable
+      # fields for each Salesforce SObject Type.
+      #
+      # Rturns a Hash.
```
Comment (rdunlop) [inline lib/restforce/db/field_processor.rb]: typo

File: lib/restforce/db/field_processor.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      end
+
+      # Internal: Clear out the global field cache.
+      #
+      # Returns nothing.
+      def self.reset
+        @field_cache = {}
+      end
+
+      # Pubic: Get a restricted version of the passed attributes Hash, with
```
Comment (rdunlop) [inline lib/restforce/db/field_processor.rb]: typo `Public`

**Comment by rdunlop** [comment]
:+1:

File: lib/restforce/db/accumulator.rb [hunk 3]
```diff
@@ -44,6 +44,19 @@ def current(comparison)
         end
       end
 
+      # Public: Do the canonical attributes stored in this Accumulator differ
+      # from those in the passed comparison Hash?
+      #
+      # comparison - A Hash mapping of attributes to values.
+      #
+      # Returns a Boolean.
+      def changed?(comparison)
```
Comment (danhodos) [inline lib/restforce/db/accumulator.rb]: I dig this change to not build up the entire Hash.

File: lib/restforce/db/synchronizer.rb [hunk 4]
```diff
@@ -49,10 +49,11 @@ def run(changes)
       #
       # Returns nothing.
       def update(instance, accumulator)
+        return unless accumulator.changed?(instance.attributes)
```
Comment (danhodos) [inline lib/restforce/db/synchronizer.rb]: In seeing this in use, I wonder if the method should be `differs_from?` or somesuch, as it reads a little more smoothly.

File: test/lib/restforce/db/field_processor_test.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  describe "#process" do
+    let(:attributes) do
+      {
+        "Name" => "This field can be updated.",
+        "Id"   => "But... but... this field is read-only!",
+      }
+    end
+    let(:dummy_client) do
+      Object.new.tap do |client|
+
```
Comment (danhodos) [inline test/lib/restforce/db/field_processor_test.rb]: This extra blank line is weird, or the lack of one after line 27 is.

**Comment by danhodos** [comment]
@ahorner Any more plans for `FieldProcessor` right now? I wondered because `FieldProcessor` is a fairly generic name; however, if you're anticipating it growing to include more general sort of processing in the future then it makes sense.

**Comment by danhodos** [comment]
Had a couple quibbles, but otherwise :+1:

File: lib/restforce/db/synchronizer.rb (hunk 4 above)
Comment (ahorner) [inline lib/restforce/db/synchronizer.rb]: I feel like `#changed?` conveys more information for this system, regarding what has actually happened to the attributes.

**Comment by ahorner** [comment]
@danhodos: I don't have any immediate plans, but there's enough metadata returned by Salesforce that using a more specific class name seemed pointless; I wanted this to be sort of a black box to the rest of the system.

**Comment by danhodos** [comment]
> I wanted this to be sort of a black box to the rest of the system.

Heh heh, I just don't want it to be a black-box to future-us. I'm okay with leaving it as it (it's not bad, just generic, kind of like most pop music).

**Pull request merged**